### ISSUE TITLE
User-supplied fail_calc for tests

### ISSUE THREAD

**Issue by jtcohen6** [fishtown-analytics/dbt#3321]
### Describe the feature

Today, the `'test'` materialization hard-codes `count(*)` as the way to calculate failures from a test:

https://github.com/fishtown-analytics/dbt/blob/26fb58bd1b08218781b182918f5ed4dec3f735d9/core/dbt/include/global_project/macros/materializations/test.sql#L4-L7

I think that makes sense 90% of the time in the general case, but we want to give users the ability to customize the "failure calculation" if they'd like it to be something other than `count(*)`. This is important for backward compatibility, since schema tests could previously calculate and return whatever numeric value they wanted. In the wild, this could be as simple as `sum(column)` instead of `count(*)`, or it could be as complex as the [`dbt_utils.equality`](https://github.com/fishtown-analytics/dbt-utils/blob/489a44bb7ebad1a6d9ba762fdcd7dc6a9fea7e0d/macros/schema_tests/equality.sql#L68-L73) test:

```sql
select count(*) from unioned) +
        (select abs(
            (select count(*) from a_minus_b) -
            (select count(*) from b_minus_a)
            )
```

I'm hopeful this is quite straightforward to implement—it's just a matter of pulling in the `fail_calc` and templating it into the materialization.

### Questions

- Should `fail_calc` be a test _config_ or a test _property_? I lean toward _property_, since I think this is an essential component of the test definition and less like something that wants to be set for many different types of tests at once, e.g. from `dbt_project.yml`. (In a post-#2401 world, this is hopefully a less meaningful distinction!)
- Could this have potentially strange interactions with `%` values of `warn_if` / `error_if` (#3258)? Yes! I don't think we need to solve for every edge case there now.

**Comment by jtcohen6** [comment]
Update: We're going to make this a _test config_, following the pattern sketched out in #3258.

I don't think it makes a _ton_ of sense for users to override the default configs set by the generic test definition, but hey, they'll be able to if they want to. I could even see this being compelling—let's say you want the `unique` test to calculate failure as _the number of original rows containing a duplicate_, rather than _the length of the set of duplicate values_.

**Comment by jtcohen6** [comment]
Resolved by #3392
